### ISSUE TITLE
Generating additional styles from variables/config

### ISSUE THREAD

**Issue by missmatsuko** [bigfishtv/turret#36]
I'm taking a look at different helper/utility CSS libraries, and I'm having a tough time picking one!

I like that turretcss uses plain CSS with CSS variables, and the class names are generally pretty straight-forward.

One thing I wish this had, though, is the generation of more styles from a config.

For example, some sites have way more than 3 theme colours (primary/secondary/tertiary). It would be great to be able to add more variables to `_color.css` and have all the color, border, background, etc. styles generated for those extra variables.

**Comment by scottdejonge** [comment]
You've read my mind!

I've been toying with this idea for a while. Especially seeing the flexibly and success of [TailwindCSS](https://tailwindcss.com).

I would love to abstract the opinionated logic (colors, sizes xxs-xxl etc.) to be compiled from config so that there can be greater flexibility.

I'm not sure what the best solution is here. I want to avoid technical debt and maintain a simplistic workflow (currently contained to PostCSS alone).

In terms of atomic/utility class CSS libraries I would suggest picking the one that best suits your problem space. [TailwindCSS](https://tailwindcss.com) is popular, and has more documentation and support with a dedicated [Slack channel](https://tailwindcss.slack.com/). Not to mention it's creator @adamwathan's work with [Refactoring UI](https://refactoringui.com/) is amazing and super helpful!

TurretCSS is used by our internal team for the websites we develop for our clients. Updates and support is slower and might be more opinionated than something more widely adopted.

Also worth mentioning [Tachyons](http://tachyons.io/) ([Github](https://github.com/tachyons-css/tachyons/))by @mrmrs and [Basscss](https://basscss.com/) ([Github](https://github.com/basscss/basscss/)) by @jxnblk which again are more popular and widely used with greater support.

I hope that helps.

**Comment by missmatsuko** [comment]
> I'm not sure what the best solution is here. I want to avoid technical debt and maintain a simplistic workflow (currently contained to PostCSS alone).

Hmm yes, I was wondering if a PostCSS plugin could read CSS variables and generate the styles. I don't know much about writing PostCSS plugins (yet) though. 

I might end up going with Tailwind anyways, but I much prefer turret's class names!

**Comment by scottdejonge** [comment]
Yeah I would love to have a really simple json file that could be read and then generate the utility classes. I also don't know much about writing PostCSS plugins, I've bene leveraging other's work in that area to keep it generic and unopinionated.

I fully support going with Tailwind. I'm tempted to make the switch myself sometimes... but I also prefer the TurretCSS classnames, at at this point it's pretty much muscle memory.

**Comment by gavinmcfarland** [comment]
I haven't had the pleasure of using TurretCSS yet but it looks like a lot of it is based on custom properties. I'm familiar with creating Postcss plugins and it wouldn't be too hard to create a config which generates new custom properties or classes.

The most time consuming aspect is thinking of the architecture of what the config would look like and how that would map to your CSS.

There is a great template you can use to [create a Postcss plugin](https://github.com/csstools/create-postcss-plugin) made by Jonathan Neal. I'm a bit busy myself, but if you get stuck or need some help I'll do my best.

Another way would be to utilise existing Postcss plugins like mixins to generate styles from variables in a stylesheet which acts as your config.

**Comment by scottdejonge** [comment]
Thanks for your insight @limitlessloop!

I've been exploring the possibility of `json` configuration as part of the [`importFrom`](https://github.com/csstools/postcss-preset-env#importfrom) functionaility in [PostCSS Preset Env](https://preset-env.cssdb.org/).

As all configuration options are provided as custom prosperities it should be possible to then map those same values to a config file containing the key/value pairs.

The main issue I'm faced with here is how to define config. Since turretcss aims to be modern CSS only it feels conflicting to move config into `json` therefore requiring a build step.

On the other hand, having both `json` and `css` (or `json` that compiles to the `css` variable definitions which is committed to the repo) may provide the best of both worlds.

I'm ultimately hoping to provide the ability for javascript intervention and compilation if required, whilst still maintaining the CSS only methodology.

I will continue to explore options, and keep this issue open to any and all feedback.

**Comment by gavinmcfarland** [comment]
Yes I think I see where you're coming from. In that case I think you're probably right. For the time being maybe it would be best to keep the configuration options inside CSS until you can think of how you'll map the JSON to CSS. Others can always integrate their own tool to convert JSON to CSS, like using Theo or Style Dictionary.

**Comment by ckizer** [comment]
I'm confused on CSS Variable support. I linked to a precompiled version of turrent like so:
@import turretcss.min.css

Directly below that I tried redefining some variables and it doesn't seem to work. Is turret using try CSS variables that I can override?

```
CSS
:root
    --border-radius: 1em
    --primary: red !important

HTML
    <div class="background-primary">Still shows blue and not red</div>
```

This doesn't override the primary color or border radius. Do the css variables overrides only work if I download and compile the CSS

**Comment by adamroyle** [comment]
> I'm confused on CSS Variable support. I linked to a precompiled version of turrent like so:
> @import turretcss.min.css

Importing a pre-compiled css file will have the CSS variables already replaced with their values.

You'll need to import the uncompiled css first, and then override the values.

```
@import 'turretcss/turret/turret';

:root {
   --border-radius: 1em;
   --primary: red;
}
```

**Comment by ckizer** [comment]
Thanks Adam. I was hoping to avoid including a large library, but rather a single file. I've included it in my static HTML project as you suggested and it's working great.

I really like the library, but wouldn't it make sense for the default base font-size to affect all the other elements. I assumed that changing `  --font-size-m: 1rem;` to `--font-size-m: 2rem;` would automatically scale the rest of the type.

I'm guessing there's a logical reason? It seems like using `rem` font sizes would be the perfect way to scale the headings and the display headings? Maybe I'm missing the best way to do it?

I'm considering writing this to make it happen

```
  --scale: 1.333;
  /* Don't change the lines below, you only need to change the --scale above. */
  --size-l-1: calc(1rem * var(--scale));
  --size-l-2: calc(var(--size-l-1) * var(--scale));
  --size-l-3: calc(var(--size-l-2) * var(--scale));
  --size-l-4: calc(var(--size-l-3) * var(--scale));
  --size-l-5: calc(var(--size-l-4) * var(--scale));
  --size-l-6: calc(var(--size-l-5) * var(--scale));
  --size-l-7: calc(var(--size-l-6) * var(--scale));
  --size-s-1: calc(1rem / var(--scale));
  --size-s-2: calc(var(--size-s-1) / var(--scale));
  --size-s-3: calc(var(--size-s-2) / var(--scale));
```

Thanks for the library. The class names are logical and consistent :-)

PS: Maybe this is because I'm using CodeKit to compile and pull the css in rather than the NPM libraries

**Comment by adamroyle** [comment]
@ckizer

This conversation is going a bit off-topic for this issue, however in short it already is relative by the fact it uses rems, so you can change the default `--root-size: 16px;` to something larger. Note that this will also affect space/padding/margin. We have found 16px to be a good default for most of the sites we create so we generally stick with that. 

Something else to understand is this library is designed to be used with PostCSS and compiled to a single css file. Turret was developed before browsers had support for custom properties. We generally copy the `_base.css`, `_color.css` and other property declaration files into our projects, include them after the main Turret import and modify them as required.

**Comment by ckizer** [comment]
@adamroyle Understood. Thanks. I'll move my other questions to a different thread. Thanks for the clarification. I'm going to dive in head first ;-)
